docs/internals.rst

```diff
@@ -0,0 +1,314 @@
+each time.
+
+The archive metadata does not contain the file items directly. Only
+references to other objects that contain that data. An archive is an
+object that contain metadata:
+
+* version
+* name
+* items list
+* cmdline
+* hostname
+* username
+* time
+
+Each item represents a file or directory or
+symlink is stored as a ``item`` dictionnary that contains:
+
+* path
+* list of chunks
+* user
+* group
+* uid
+* gid
+* mode (item type + permissions)
+* source (for links)
+* rdev (for devices)
+* mtime
+* xattrs
+* acl
+* bsdfiles
+
+``ctime`` (change time) is not stored because there is no API to set
+it and it is reset every time an inode's metadata is changed.
+
+All items are serialized using msgpack and the resulting byte stream
+is fed into the same chunker used for regular file data and turned
+into deduplicated chunks. The reference to these chunks is then added
+to the archvive metadata. This allows the archive to store many files,
+beyond the ``MAX_OBJECT_SIZE`` barrier of 20MB.
+
+A chunk is an object as well, of course, and its id is the hash of its
+(unencrypted and uncompressed) content.
+
+Hints are stored in a file (``repo/hints``) and contain:
+
+* version
+* list of segments
+* compact
+
+Chunks
+------
+
+|project_name| uses a rolling checksum with Buzhash_ algorithm, with
+window size of 4095 bytes, with a minimum of 1024, and triggers when
+the last 16 bits of the checksum are null, producing chunks of 64kB on
+average. All these parameters are fixed. The buzhash table is altered
+by XORing it with a seed randomly generated once for the archive, and
+stored encrypted in the keyfile.
+
+Indexes
+-------
+
+There are two main indexes: the chunk lookup index and the repository
+index. There is also the file chunk cache.
+
+The chunk lookup index is stored in ``cache/chunk`` and is indexed on
+the ``chunk hash``. It contains:
+
+* reference count
+* size
+* ciphered size
+
+The repository index is stored in ``repo/index.%d`` and is also
+indexed on ``chunk hash`` and contains:
+
+* segment
+* offset
+
+The repository index files are random access but those files can be
+recreated if damaged or lost using ``check --repair``.
+
+Both indexes are stored as hash tables, directly mapped in memory from
+the file content, with only one slot per bucket, but that spreads the
+collisions to the following buckets. As a consequence the hash is just
+a start position for a linear search, and if the element is not in the
+table the index is linearly crossed until an empty bucket is
+found. When the table is full at 90% its size is doubled, when it's
+empty at 25% its size is halfed. So operations on it have a variable
+complexity between constant and linear with low factor, and memory
+overhead varies between 10% and 300%.
+
+The file chunk cache is stored in ``cache/files`` and is indexed on
+the ``file path hash`` and contains:
+
+* age
+* inode number
+* size
+* mtime_ns
+* chunks hashes
```

Comment:
i was referring to a potential grammar error / typo there. not sure though, therefore the "?"